### DISCUSSION TITLE
UM FeatherS3 becomes catatonic after entering second deep sleep...

### DISCUSSION THREAD

**Discussion by JDRBoston** [micropython/micropython#12509] [ESP32] [answered]
After the ESP-S3 enters a second deepsleep, it becomes unresponsive. I am stumped and would greatly appreciate your advice on how to further troubleshoot/fix this problem. I could not find a solution on-line.

I am using Unexpected maker FeatherS3 version P7 flashed with UM_FEATHERS3-20230426-v1.20.0.bin. The interpreter works otherwise. Here's what I've tried so far. I loaded the troubleshooting script below as 'main.py' onto the device. I then reset it, allowed it to cycle through the script, flashing code of '0' as expected and then connected to its REPL interface. I started the first deepsleep cycle by executing 'import machine; machine.deepsleep(1000)'. As expected, the device closes the USB connection, goes to sleep for 1 second, wakes up - flashing 'a', recognizes that it woke from deepsleep by telegraphing Morse code for '5', and then goes to sleep. But it won't wake up! Accordingly, the device does not show up on my USB bus and ignores desperate attempts to reconnect to its USB/REPL using the last port designation. Restarting the device fires the main.py script--ending with the power on reset code being flashed--and permits a connection to REPL.

Thank you for your help! J

`import machine
from machine import Pin, deepsleep
import time

morse_numbers_dict = {
    'a': '.-', 0: '-----', 1: '.----', 2: '..---', 3: '...--', 4: '....-',
    5: '.....', 6: '-....', 7: '--...', 8: '---..', 9: '----.'
}

# Instantiate FeatherS3 blue LED
blue_led = Pin(13, Pin.OUT); blue_led.value(0)

def dot():
    blue_led.value(1)
    time.sleep(0.2)
    blue_led.value(0)
    time.sleep(0.2)
    
def dash():
    blue_led.value(1)
    time.sleep(0.6)
    blue_led.value(0)
    time.sleep(0.2)

def morse_code(code):
    for symbol in morse_numbers_dict[code]:
        if symbol == '.': dot()
        if symbol == '-': dash()
            

# show life
for _ in range(2):
    morse_code('a')
    time.sleep(1)
    
time.sleep(3)

# signal wakeup cause
if machine.reset_cause() == machine.DEEPSLEEP_RESET: # = 5
    morse_code(5)
    deep_sleep(1000)

if machine.reset_cause() == machine.HARD_RESET: # = 6
    morse_code(6)

if machine.reset_cause() == machine.PWRON_RESET: # = 0
    morse_code(0)

if machine.reset_cause() == machine.WDT_RESET: # = 1
    morse_code(1)

if machine.reset_cause() == machine.SOFT_RESET: # = 4
    morse_code(4)

print('end of program')`

**Comment by davefes** [comment, marked as answer]
> deep_sleep(1000)

-> ` deepsleep(1000)`

Are you sure it didn't throw an error?

**Reply by glenn20** [reply]
+1 for dave's reply. Although, I don't understand why it would have locked up and not dropped to the repl after generating an exception.

Also, please read the "please read before posting" message at the top of https://github.com/orgs/micropython/discussions. That will explain how to make sure you code is formatted correctly when you post, which makes the job of people trying to help a lot easier.

**Reply by JDRBoston** [reply]
Thank you, very much, Dave I missed that error. Fixing it solved the problem - the FeatherS3 is waking up from deepsleep as expected. Thank you Glenn - I now understand how to enter the scripts so that they are easier to read.

**Comment by peterhinch** [comment]
Another +1. Is there also a possible USB issue? A deepsleep will cause USB to fail. On wakeup the S3 will re-initialise its USB port however I'm not sure if the PC is guaranteed to respond. This may depend on hardware and OS.

**Reply by JDRBoston** [reply]
Hi Peter, I'm using a Mac and programming the S3 using Terminal and ampy. Terminal shuts down the connection when the S3 sleeps and does not automatically re-establish a connection when the S3 wakes up. This made troubleshooting difficult because I didn't see an error. For future troubleshooting, is there a way to save the REPL error output to a file on the S3. I looked for some documentation and could not find it.

**Reply by peterhinch** [reply]
I don't know of a way. The ideal is to have a way of mapping the REPL onto a UART, but few platforms can do that (STM can). The way I debug micropower code is to develop using `time.sleep`, only changing to `deepsleep` when it's basically working.

**Comment by andrewleech** [comment]
I've dealt with a similar issue on stm that could be related: https://github.com/micropython/micropython/pull/8304

If you can get the board powered by a separate supply it would be interesting to know if unplugging & replugging the USB gets it going again

**Reply by peterhinch** [reply]
FWIW I develop micropower applications on the assumption that USB will not work. On some platforms the REPL can be redirected to a serial port. Another approach is to replace `machine.deepsleep` with `time.sleep` while debugging.

**Comment by JDRBoston** [comment]
Thank you, everyone! Dave picked up my mistake. Here's some S3 working deepsleep code in case it is helpful to anyone:

```py
import machine
from machine import Pin, deepsleep
import time

morse_numbers_dict = {
    'a': '.-', 0: '-----', 1: '.----', 2: '..---', 3: '...--', 4: '....-',
    5: '.....', 6: '-....', 7: '--...', 8: '---..', 9: '----.'
}

# Instantiate FeatherS3 blue LED
blue_led = Pin(13, Pin.OUT); blue_led.value(0)

def dot():
    blue_led.value(1)
    time.sleep(0.2)
    blue_led.value(0)
    time.sleep(0.2)
    
def dash():
    blue_led.value(1)
    time.sleep(0.6)
    blue_led.value(0)
    time.sleep(0.2)

def morse_code(code):
    for symbol in morse_numbers_dict[code]:
        if symbol == '.': dot()
        if symbol == '-': dash()
            

# show life
for _ in range(2):
    morse_code('a')
    time.sleep(1)
    
time.sleep(3)

# signal wakeup cause
if machine.reset_cause() == machine.DEEPSLEEP_RESET: # = 5
    morse_code(5)
    deepsleep(1000)

if machine.reset_cause() == machine.HARD_RESET: # = 6
    morse_code(6)

if machine.reset_cause() == machine.PWRON_RESET: # = 0
    morse_code(0)

if machine.reset_cause() == machine.WDT_RESET: # = 1
    morse_code(1)

if machine.reset_cause() == machine.SOFT_RESET: # = 4
    morse_code(4)

print('end of program')
```
